src/processor.cpp

```diff
@@ -0,0 +1,401 @@
+
+    std::string full_message;
+    std::string empty_message;
+    std::string start_message;
+    std::string processing_message;
+    std::string fail_message;
+
+    bool was_loaded;
+
+    void load( JsonObject &jo, const std::string &src );
+    void check() const;
+    void reset();
+
+};
+
+processor_id get_processor_id( const tripoint &pos )
+{
+    return processor_id( g->m.furn( pos ).id().str() );
+}
+
+generic_factory<processor_data> processors_data( "processor type", "id" );
+
+}
+
+void processor_data::load( JsonObject &jo, const std::string & )
+{
+    mandatory( jo, was_loaded, "next_type", next_type );
+    mandatory( jo, was_loaded, "is_processing", is_processing );
+    mandatory( jo, was_loaded, "fuel_type", fuel_type );
+    mandatory( jo, was_loaded, "processes", processes, string_id_reader<process_id> {} );
+
+    if( !was_loaded || jo.has_member( "messages" ) ) {
+        JsonObject messages_obj = jo.get_object( "messages" );
+
+        optional( messages_obj, was_loaded, "empty", empty_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "full", full_message, translated_string_reader );
+        //optional( messages_obj, was_loaded, "prompt", prompt_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "start", start_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "processing", processing_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "finish", finish_message, translated_string_reader );
+        optional( messages_obj, was_loaded, "fail", fail_message, translated_string_reader );
+    }
+}
+
+void processor_data::check() const
+{
+    static const iexamine_function converter( iexamine_function_from_string( "converter" ) );
+    const furn_str_id processor_id( id.str() );
+
+    if( !processor_id.is_valid() ) {
+        debugmsg( "Furniture \"%s\" have no processor of the same name.", id.c_str() );
+    } else if( processor_id->examine != converter ) {
+        debugmsg( "Furniture \"%s\" can't act as a processor because the assigned iexamine function is not converter, but it has \"%s\" associated.",
+                  processor_tid.c_str(), id.c_str() );
+    }
+    for( const auto &elem : processes ) {
+        if( !elem.is_valid() ) {
+            debugmsg( "Invalid process \"%s\" in \"%s\".", elem.c_str(), id.c_str() );
+        }
+    }
+}
+
+void processor_data::load( JsonObject &jo, const std::string &src )
+{
+    processors_data.load( jo, src );
+}
+
+void processor_data::check()
+{
+    processors_data.check();
+}
+
+void processor_data::reset()
+{
+    processors_data.reset();
+}
+
+//process namespace
+
+namespace
+{
+
+struct process_data;
+using process_id = string_id<process_data>;
+
+struct process_data {
+    process_data():
+        was_loaded( false ) {};
+
+    process_id id;
+    std::string name;
+    auto components;
+    auto output;
```

Comment:
Generally avoid auto in fact, it makes code harder to read and the really good uses of it are rather esoteric.